syntax = "proto3";

package api.v1;

import "google/protobuf/empty.proto";

option go_package = "github.com/synnaxlabs/synnax/pkg/api/grpc/v1";

service HardwareCreateTaskService {
  rpc Exec(HardwareCreateTaskRequest) returns (HardwareCreateTaskResponse);
}

service HardwareRetrieveTaskService {
  rpc Exec(HardwareRetrieveTaskRequest) returns (HardwareRetrieveTaskResponse);
}

service HardwareDeleteTaskService {
  rpc Exec(HardwareDeleteTaskRequest) returns (google.protobuf.Empty);
}

service HardwareCreateRackService {
  rpc Exec(HardwareCreateRackRequest) returns (HardwareCreateRackResponse);
}

service HardwareRetrieveRackService {
  rpc Exec(HardwareRetrieveRackRequest) returns (HardwareRetrieveRackResponse);
}

service HardwareDeleteRackService {
  rpc Exec(HardwareDeleteRackRequest) returns (google.protobuf.Empty);
}

service HardwareCreateDeviceService {
  rpc Exec(HardwareCreateDeviceRequest) returns (HardwareCreateDeviceResponse);
}

service HardwareRetrieveDeviceService {
  rpc Exec(HardwareRetrieveDeviceRequest) returns (HardwareRetrieveDeviceResponse);
}

service HardwareDeleteDeviceService {
  rpc Exec(HardwareDeleteDeviceRequest) returns (google.protobuf.Empty);
}

message Rack {
  uint32 key = 1;
  string name = 2;
}

message HardwareCreateRackRequest {
  repeated Rack racks = 1;
}

message HardwareCreateRackResponse {
  repeated Rack racks = 1;
}

message HardwareRetrieveRackRequest {
  repeated uint32 keys = 1;
  repeated string names = 2;
}

message HardwareRetrieveRackResponse {
  repeated Rack racks = 1;
}

message HardwareDeleteRackRequest {
  repeated uint32 keys = 1;
}

message Task {
  uint64 key = 1;
  string name = 2;
  string type = 3;
  string config = 4;
  bool internal = 5;
  bool snapshot = 6;
}

message HardwareCreateTaskRequest {
  repeated Task tasks = 1;
}

message HardwareCreateTaskResponse {
  repeated Task tasks = 1;
}

message HardwareRetrieveTaskRequest {
  uint32 rack = 1;
  repeated uint64 keys = 2;
  repeated string names = 3;
  repeated string types = 4;
}

message HardwareRetrieveTaskResponse {
  repeated Task tasks = 1;
}

message HardwareDeleteTaskRequest {
  repeated uint64 keys = 1;
}

message Device {
  string key = 1;
  string name = 2;
  uint32 rack = 3;
  string location = 4;
  string make = 6;
  string model = 7;
  string properties = 8;
  bool configured = 9;
}

message HardwareCreateDeviceRequest {
  repeated Device devices = 1;
}

message HardwareCreateDeviceResponse {
  repeated Device devices = 1;
}

message HardwareRetrieveDeviceRequest {
  repeated string keys = 1;
  repeated string names = 2;
  repeated string makes = 3;
  repeated string models = 4;
  repeated string locations = 5;
  repeated uint32 racks = 6;
  string search = 7;
  uint32 limit = 8;
  uint32 offset = 9;
}

message HardwareRetrieveDeviceResponse {
  repeated Device devices = 1;
}

message HardwareDeleteDeviceRequest {
  repeated string keys = 1;
}
